Add gitlab ci

pbaughman opened the pull request:
~~Depends on https://github.com/ApexAI/apex_rostest/pull/6.  Rebase this after it's merged so there won't be 77 files changed~~

  - Build and test apex_launchtest only
  - Build and test everything (apex_launchtest + apex_launchtest_ros)
  - Coverage report HTML is uploaded as job artifact

Take a look at an example [HTML report generated from this job](https://apexai.gitlab.io/-/apex_rostest/-/jobs/164164150/artifacts/htmlcov/index.html)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/apexai/apex_rostest/8)
<!-- Reviewable:end -->

Commit:
Enable gitlab CI

  - Build and test apex_launchtest only
  - Build and test everything (apex_launchtest + apex_launchtest_ros)
  - Coverage report

pbaughman commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,72 @@
+
+build_launchtest:
+  stage: build
+  image: osrf/ros2:nightly
```
I'd prefer to have this run on osrf/ros2:bouncy or osrf/ros2:crystal but it doesn't look like those docker images are [available anymore](https://hub.docker.com/r/osrf/ros2/tags).  I'm guessing they were in the past, because ApexAI [performance test package](https://github.com/ApexAI/performance_test/blob/master/.gitlab-ci.yml#L7) looks like it used it before

dejanpan replied:
Try `FROM osrf/ros:crystal-desktop`

pbaughman replied:
This works, except that we depend on a slightly newer version of launch than what's in crystal.  For now, I'll leave this on the 'nightly'

pbaughman commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,72 @@
+
+coverage:
+  stage: report
+  image: osrf/ros2:nightly
```
This could use a smaller docker image to speed up the job - it just needs to have pip3 and python3 installed to work.  I don't know who, if anybody, publishes such an image.

Commit:
Speed up the 'coverage' job

pbaughman commented:
@left4taco I'm done touching this - you can review now

t0ny-peng commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,74 @@
+  - report
+
+before_script:
+  - sh -c 'echo "deb [arch=amd64,arm64] http://repo.ros2.org/ubuntu/main `lsb_release -cs` main" > /etc/apt/sources.list.d/ros2-latest.list'
```
@pbaughman I'm curious. What is the point of running the command explicitly with `sh` here?

pbaughman replied:
It's probably unnecessary.  I copy/pasted [from here](https://index.ros.org/doc/ros2/Installation/Linux-Install-Debians/)

> sudo sh -c 'echo "deb [arch=amd64,arm64] http://packages.ros.org/ros2/ubuntu `lsb_release -cs` main" > /etc/apt/sources.list.d/ros2-latest.list'

The sh -c was probably just to make the whole thing happen under sudo.  The docker image didn't like sudo (i'm already root in the docker image) so I got rid of it, but not the sh -c part.  I will remove it

t0ny-peng replied:
@pbaughman Oh since you mentioned `sudo` I understand why they use `sh -c`. 

If you `sudo echo xxxxx > /etc/qqqqq`, it'll fail as `sodu` only works for the part before `>`, the output redirection. The rest part still doesn't have enough permission. This is a bash feature.

t0ny-peng commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,74 @@
+    - colcon build
+  artifacts:
+    paths:
+      - .
```
Nitpick: will it be better to upload only the `build` and `install` folders? If you upload everything under `.`, `.git` will also be uploaded and then override the same folder in job `test_all`. Seems a potential bug to me.

pbaughman replied:
Yes, I can do that

t0ny-peng commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,74 @@
+  artifacts:
+    when: always
+    paths:
+      - .
```
Same here. Maybe only upload the part that we care, i.e., log?

pbaughman replied:
I can't change this one - the .coverage files are generated in the package roots and I need those to be passed along.  Is there a good way to tar up the .coverage files and then untar them back into the same directories without having to explicitly list apex_launchtest, apex_launchtest_cmake, and apex_launchtest_ros?  I'm concerned that in the future if another package is added, it will get missed for coverage

t0ny-peng replied:
You can use `--cov-report` to specify the output directory. More info can be found here https://pytest-cov.readthedocs.io/en/latest/reporting.html

That being said, it's up to you to decide whether or not to use it. 

---

But why not move the whole coverage job into `test_all`?

pbaughman replied:
re: the last part - I don't want to generate coverage unless everything passes.

t0ny-peng commented on .gitlab-ci.yml:
```diff
@@ -0,0 +1,74 @@
+  script:
+    - pip3 install mock
+    - colcon test --pytest-args "--cov"
+    - colcon test-result --verbose
```
You can now use `colcon test --return-code-on-test-failure ...` to do the same thing with `colcon test-result`

pbaughman replied:
I'll try it, but the --verbose is important too - it will print where the test failure was without having to search around and without having to dig through the log artifacts

t0ny-peng commented:
@pbaughman I've done my review. Please address my comments and then ping me here. Thanks.

pbaughman commented:
@left4taco Replied to comments.  The only tricky one is the artifacts from the test_all job.  I need the .coverage files to persist over to the 'coverage' job and they live in alongside the source files

t0ny-peng commented:
@pbaughman I've added more to the review. Please push your change and ping me. Thanks.

Commit:
Address Hao's PR comments

  - Git rid of extra sh -c (left over from a copy/paste)
  - Build jobs only save 'build' and 'install' artifacts
  - test_all job saves 'build', 'install', 'log', and globs for .coverage files

pbaughman commented:
@left4taco Ok, I think that takes care of everything

t0ny-peng commented:
@pbaughman All look good to me. Please let Dejan to approve this PR and then you can merge. Thanks.

dejanpan reviewed (approved):
LGTM in that I do not see anything obvious wrong. 
@left4taco gave an expert review.
